slick/src/main/scala/slick/util/ManagedArrayBlockingQueue.scala

```diff
@@ -1,87 +1,71 @@
 package slick.util
 
-import java.util.concurrent.{ArrayBlockingQueue, TimeUnit, BlockingQueue}
+import java.util.concurrent.{TimeUnit, BlockingQueue}
 import java.util.concurrent.locks._
-import java.util._
+import java.util
+
+import slick.util.AsyncExecutor.{LowPriority, MediumPriority, HighPriority, Priority}
 
 /** A simplified copy of `java.util.concurrent.ArrayBlockingQueue` with additional logic for
   * temporarily rejecting elements based on the current size. All features of the original
   * ArrayBlockingQueue have been ported, except the mutation methods of the iterator. See
   * `java.util.concurrent.ArrayBlockingQueue` for documentation. */
-abstract class ManagedArrayBlockingQueue[E >: Null <: AnyRef](capacity: Int, fair: Boolean = false) extends AbstractQueue[E] with BlockingQueue[E] { self =>
+abstract class ManagedArrayBlockingQueue[E >: Null <: AnyRef](capacity: Int, fair: Boolean = false)
+  extends util.AbstractQueue[E]
+  with BlockingQueue[E]
+  with Pausable { self =>
 
-  /** Determine if the item should be accepted at the current time. */
-  protected[this] def accept(item: E, size: Int): Boolean
+  protected[this] def priority(item: E): Priority
 
-  private[this] val items = new Array[AnyRef](capacity)
   private[this] val lock = new ReentrantLock(fair)
   private[this] val notEmpty = lock.newCondition
   private[this] val notFull = lock.newCondition
-  private[this] var takeIndex, putIndex, count = 0
 
   private[this] def checkNotNull(v: AnyRef): Unit = if (v == null) throw new NullPointerException
 
-  private[this] def inc(i: Int): Int = if(i+1 == items.length) 0 else i+1
-
-  private[this] def dec(i: Int): Int = (if(i == 0) items.length else i) - 1
+  private[this] val itemQueue = new InternalArrayQueue[E](2*capacity)
```

Comment:
It is the same as for the original ManagedArrayBlockingQueue. You need twice the capacity for the low/medium priority queue, because you always need to be able to schedule medium priority items on the queue. These medium priority are continuations of the low priority items. @szeiger can explain it better